java/arcs/core/data/expression/Evaluator.kt

```diff
@@ -164,42 +168,69 @@ sealed class GlobalFunction(val name: String) {
       toSequence<Any>(args[0]).asIterable().union(
         toSequence<Any>(args[1]).asIterable()
       ).asSequence()
+
+    override fun inferredType(evaluator: TypeEvaluator, args: List<InferredType>) =
+      InferredType.UnionType(args.toSet())
   }
 
   /** Find the maximum of a [Sequence]. */
   object Max : GlobalFunction("max") {
     override fun invoke(evaluator: ExpressionEvaluator, args: List<Any?>) =
       toSequence<Comparable<Comparable<*>>>(args[0]).max()
+
+    override fun inferredType(evaluator: TypeEvaluator, args: List<InferredType>) =
+      widen(*args.toTypedArray())
   }
 
   /** Find the minimum of a [Sequence]. */
   object Min : GlobalFunction("min") {
     override fun invoke(evaluator: ExpressionEvaluator, args: List<Any?>) =
       toSequence<Comparable<Comparable<*>>>(args[0]).min()
+
+    override fun inferredType(evaluator: TypeEvaluator, args: List<InferredType>) =
+      widen(*args.toTypedArray())
   }
 
   /** Find the average of a [Sequence]. */
   object Average : GlobalFunction("average") {
     override fun invoke(evaluator: ExpressionEvaluator, args: List<Any?>) =
       toSequence<Int>(args[0]).average()
+
+    override fun inferredType(evaluator: TypeEvaluator, args: List<InferredType>) =
+      InferredType.Primitive.DoubleType
   }
 
   /** Find the average of a [Sequence]. */
   object Sum : GlobalFunction("sum") {
     override fun invoke(evaluator: ExpressionEvaluator, args: List<Any?>) =
       toSequence<Int>(args[0]).sum()
+
+    override fun inferredType(evaluator: TypeEvaluator, args: List<InferredType>) =
+      widen(*args.toTypedArray())
   }
 
   /** Count the number of elements in a [Sequence]. */
   object Count : GlobalFunction("count") {
     override fun invoke(evaluator: ExpressionEvaluator, args: List<Any?>) =
       toSequence<Any>(args[0]).count()
+
+    override fun inferredType(evaluator: TypeEvaluator, args: List<InferredType>) =
+      InferredType.Primitive.IntType
   }
 
   /** Return the first item of a [Sequence]. */
   object First : GlobalFunction("first") {
     override fun invoke(evaluator: ExpressionEvaluator, args: List<Any?>) =
       toSequence<Any>(args[0]).firstOrNull()
+
+    override fun inferredType(evaluator: TypeEvaluator, args: List<InferredType>) =
+      InferredType.UnionType(
+        InferredType.Primitive.NullType,
+        when (val it = args[0]) {
```

Comment:
`it` is not that descriptive of a name.

Reply:
What if `args` is an empty list?

Reply:
If it's two cases, maybe an `if` statement makes sense.

Reply:
Its not allowed to take zero args and the parser will enforce that at some point in the future. The TS parser already has the arity of functions encoded, but it's not used.